src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs

```diff
@@ -0,0 +1,179 @@
+        {
+            await foreach (var job in RegisterImagesInternalAsync(identifier, false, true))
+            {
+                yield return job;
+            }
+        }
+
+        public async Task<DlcsIngestJob[]> RegisterImagesAsync(string identifier, bool useInitialOrigin = false)
+        {
+            // Can this not return an array? Why does it need to?
+            var jobs = new List<DlcsIngestJob>();
+            await foreach (var job in RegisterImagesInternalAsync(identifier, useInitialOrigin, false))
+            {
+                jobs.Add(job);
+            }
+            return jobs.ToArray();
+        }
+
+        private async IAsyncEnumerable<DlcsIngestJob> RegisterImagesInternalAsync(string identifier, bool useInitialOrigin, bool immediateStart)
+        {
+            // TODO - The unstarted old jobs are cleaned out in AddNewJob(..)
+            // But wouldn't it be better to clean them all out here? Before creating the set of new
+            // jobs for this b number (will be a set of jobs if multi volume)?
+            await foreach (var manifestationInContext in _metsRepository.GetAllManifestationsInContextAsync(identifier))
+            {
+                var job = NewJob(
+                    manifestationInContext.BNumber,
+                    manifestationInContext.SequenceIndex,
+                    manifestationInContext.VolumeIdentifier,
+                    manifestationInContext.IssueIdentifier,
+                    useInitialOrigin, immediateStart);
+                AddNewJob(job);
+                yield return job;
+            }
+        }
+
+        private void AddNewJob(DlcsIngestJob job)
+        {
+            using (var trans = ddsInstrumentationContext.Database.BeginTransaction(IsolationLevel.Serializable))
+            {
+                try
+                {
+                    // remove any unstarted jobs with the same settings
+                    var existingQuery = ddsInstrumentationContext.DlcsIngestJobs.Where(j =>
+                        j.StartProcessed == null
+                        && j.Identifier == job.Identifier
+                        && j.SequenceIndex == job.SequenceIndex);
+                    if (job.VolumePart != null)
+                    {
+                        existingQuery = existingQuery.Where(j => j.VolumePart == job.VolumePart);
+                    }
+                    if (job.IssuePart != null)
+                    {
+                        existingQuery = existingQuery.Where(j => j.IssuePart == job.IssuePart);
+                    }
+                    foreach (var dlcsIngestJob in existingQuery)
+                    {
+                        ddsInstrumentationContext.DlcsIngestJobs.Remove(dlcsIngestJob);
+                    }
+                    // now add the new one
+                    ddsInstrumentationContext.DlcsIngestJobs.Add(job);
+                    ddsInstrumentationContext.SaveChanges();
```

Comment:
By default `.SaveChanges()` will run in a Transaction so shouldn't need to manually create, unless `IsolationLevel.Serializable` is needed? See: https://docs.microsoft.com/en-us/ef/core/saving/transactions#default-transaction-behavior

Reply:
I'll leave this for now but add a comment to revisit.
I can't remember why this is in a transaction.